File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/integration/IntegrationHandler.java
package io.github.arcaneplugins.levelledmobs.bukkit.integration;

import io.github.arcaneplugins.levelledmobs.bukkit.integration.impl.CitizensIntegration;
import io.github.arcaneplugins.levelledmobs.bukkit.integration.impl.NbtApiIntegration;
import io.github.arcaneplugins.levelledmobs.bukkit.integration.type.nbt.NbtProvider;
import java.util.HashSet;
import java.util.Set;
import org.bukkit.Bukkit;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@SuppressWarnings("unused")
public final class IntegrationHandler {

    private IntegrationHandler() throws IllegalAccessException {
        throw new IllegalAccessException("Illegal instantiation of utility class");
    }

    /* vars */

    private static final Set<Integration> INTEGRATIONS = new HashSet<>();

    /* methods */

    public static boolean load() {
        // TODO add check if these integrations are disabled.
        // TODO check why this returns a boolean

        if(hasRequiredPlugins("Citizens"))
            getIntegrations().add(new CitizensIntegration());

        if(hasRequiredPlugins("NBTAPI"))
            getIntegrations().add(new NbtApiIntegration());

        return true;
    }

    public static boolean hasRequiredPlugins(final String... plugins) {
        for(var plugin : plugins) {
            if(!Bukkit.getPluginManager().isPluginEnabled(plugin)) {
                return false;
            }
        }
        return true;
    }

    @Nullable
    public static NbtProvider getPrimaryNbtProvider() {
        Integration selected = null;

        for(var integration : getIntegrations()) {
            if(!integration.isEnabled())
                continue;

            if(integration instanceof NbtProvider) {
                if(selected == null) {
                    selected = integration;
                } else if(integration.getPriority().ordinal() > selected.getPriority().ordinal()) {
                    selected = integration;
                }
            }
        }

        return (NbtProvider) selected;
    }

    /* getters and setters */

    @NotNull
    public static Set<Integration> getIntegrations() { return INTEGRATIONS; }

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/integration/IntegrationHandler.kt
package io.github.arcaneplugins.levelledmobs.bukkit.integration

import io.github.arcaneplugins.levelledmobs.bukkit.integration.impl.CitizensIntegration
import io.github.arcaneplugins.levelledmobs.bukkit.integration.impl.NbtApiIntegration
import io.github.arcaneplugins.levelledmobs.bukkit.integration.type.nbt.NbtProvider
import org.bukkit.Bukkit

object IntegrationHandler {
    val integrations = mutableSetOf<Integration>()

    fun load(): Boolean{
        // TODO add check if these integrations are disabled.
        // TODO check why this returns a boolean


        // TODO add check if these integrations are disabled.
        // TODO check why this returns a boolean
        if (hasRequiredPlugins("Citizens")) integrations
            .add(CitizensIntegration())

        if (hasRequiredPlugins("NBTAPI")) integrations
            .add(NbtApiIntegration())

        return true
    }

    fun hasRequiredPlugins(plugin: String): Boolean{
        return hasRequiredPlugins(mutableListOf(plugin))
    }

    fun hasRequiredPlugins(plugins: MutableList<String>): Boolean{
        for (plugin in plugins){
            if (!Bukkit.getPluginManager().isPluginEnabled(plugin)){
                return false
            }
        }

        return true
    }

    fun getPrimaryNbtProvider(): NbtProvider?{
        var selected: Integration? = null

        for (integration in integrations){
            if (!integration.enabled) continue

            if (integration is NbtProvider){
                if (selected == null ||
                    integration.priority.ordinal > selected.priority.ordinal){
                    selected = integration
                }
            }
        }

        return if (selected == null){
            null
        } else{
            selected as NbtProvider
        }
    }
}